### ISSUE TITLE
Non-exclusive backup mode in 9.6

### ISSUE THREAD

**Issue by amitlan** [ossc-db/pg_rman#26]
What's to do here?

http://www.postgresql.org/docs/devel/static/functions-admin.html#FUNCTIONS-ADMIN-BACKUP-TABLE

**Comment by bwtakacy** [comment]
Thanks for suggesting.

I think it is useful to implement this mode in pg_rman.
By readling the hackers discussion, the non-exclusive backup mode does not create backup_label and  this feature makes PostgreSQL stuck in backup mode by pg_rman abort.

[pgsql-hackers] Updated backup APIs for non-exclusive backups
http://www.postgresql.org/message-id/flat/CABUevEziPkz6dXOS4knwppPUtDv5Vwnk=OD1nLdFze8NhF8CcA@mail.gmail.com#CABUevEziPkz6dXOS4knwppPUtDv5Vwnk=OD1nLdFze8NhF8CcA@mail.gmail.com

Am I something missing?

Anyway,  I will play with this feature in 9.6 beta1 for a while in order to know the actual usage and how to change in pg_rman side. :smiley:

**Comment by amitlan** [comment]
Sure, thanks.  I will think about it as well.

**Comment by amitlan** [comment]
Just noticed this line in 9.6 docs:

```
24.3.3. Making a Base Backup Using the Low Level API

...
Low level base backups can be made in a non-exclusive or an exclusive way.
The non-exclusive method is recommended and the exclusive one is deprecated
and will eventually be removed.
```

So, it would be necessary at some point (if not now) to change to using the non-exclusive backup API.  Also, it would be nice if it means certain usability issues of `pg_rman` itself are solved.  For example, issues like #24 should no longer happen if `pg_rman` starts using the new API if I understand it properly.

PS: I noticed that there is a new `tablespace_map` that is returned as result of `pg_stop_backup()` or is written to `$PGDATA` when using the old API.  Maybe that should have its own issue though.

**Comment by amitlan** [comment]
When reviewing this, I realized we can make our backup-from-standby process simpler.  At least in the master branch.  How? Read on.

Remember that `pg_basebackup` can be used to take backup from a standby.  For that it has to issue something like `pg_start_backup()` and `pg_stop_backup()`.  But if you try that on standby using the old API, you would get:

```
select pg_start_backup('sby-bkp-test', true);
ERROR:  recovery is in progress
HINT:  WAL control functions cannot be executed during recovery.

select pg_stop_backup();
ERROR:  recovery is in progress
HINT:  WAL control functions cannot be executed during recovery.
```

Actually, `pg_start_backup()` would try to write `backup_label` to data directory so it makes sense to prohibit the pair.

Turns out that there existed non-exclusive backup API internally which `pg_basebackup` uses since 9.2 to take standby backups.  Since it does not require writing to data directory.

And now 9.6 has the external API; so good, we can issue the non-exclusive versions of `pg_start_backup()` and `pg_stop_backup()` on standby directly.

```
select pg_start_backup('sby-bkp-test', true, false);
 pg_start_backup 
-----------------
 0/4000290
(1 row)

\! rm -r /tmp/sby-bkp && cp -r /tmp/sby /tmp/sby-bkp

select * from pg_stop_backup(false);
    lsn    |                           labelfile                           | spcmapfile 
-----------+---------------------------------------------------------------+------------
 0/4000370 | START WAL LOCATION: 0/4000290 (file 000000000000000000000004)+| 
           | CHECKPOINT LOCATION: 0/40002C8                               +| 
           | BACKUP METHOD: streamed                                      +| 
           | BACKUP FROM: standby                                         +| 
           | START TIME: 2016-05-19 16:18:47 JST                          +| 
           | LABEL: sby-bkp-test                                          +| 
           |                                                               | 
(1 row)
```

In short, we don't need to work too hard to issue `pg_start_backup()` and `pg_stop_backup()` on primary and wait for its WAL records to show up on standby and so on.  Just do the whole thing on standby.

Perhaps, we don't even need special `standby-host` and `standby-port` command line options.  Just connect to whichever server you want pg_rman to take backup from with common database connection options.

**Comment by amitlan** [comment]
So PR #27 and subsequent b37c003bc69dc1017003186f3dc8b9fd0df055da are enough for now.  Closing.  Thanks for feedback and testing.  Will open a new issue for more tests as @ooyamams suggested.
